contentcuration/contentcuration/frontend/shared/data/applyRemoteChanges.js

```diff
@@ -50,108 +57,348 @@ function transaction(change, ...args) {
-  });
-}
+class ReturnedChanges extends ChangeDispatcher {
+  /**
+   * @param {CreatedChange} change
+   * @return {Promise<void>}
+   */
+  applyCreate(change) {
+    return transaction(change, () => {
+      const table = db.table(change.table);
+      return table
+        .put(change.obj, !table.schema.primKey.keyPath ? change.key : undefined)
+        .then(() => change.obj);
+    });
+  }
 
-function applyMove(change) {
-  const resource = INDEXEDDB_RESOURCES[change.table];
-  if (!resource || !resource.tableMove) {
-    return Promise.resolve();
+  /**
+   * @param {UpdatedChange} change
+   * @return {Promise<void>}
+   */
+  applyUpdate(change) {
+    return transaction(change, () => {
+      return db
+        .table(change.table)
+        .where(':id')
+        .equals(change.key)
+        .modify(obj => applyMods(obj, change.mods));
+    });
   }
 
-  const { key, target, position } = change;
-  return resource.resolveTreeInsert(key, target, position, false, data => {
+  /**
+   * @param {DeletedChange} change
+   * @return {Promise<void>}
+   */
+  applyDelete(change) {
     return transaction(change, () => {
-      return resource.tableMove(data);
+      return db.table(change.table).delete(change.key);
     });
-  });
+  }
+
+  /**
+   * @param {MovedChange} change
+   * @return {Promise<void>}
+   */
+  applyMove(change) {
+    const resource = INDEXEDDB_RESOURCES[change.table];
+    if (!resource || !resource.tableMove) {
+      return Promise.resolve();
+    }
+
+    const { key, target, position } = change;
+    return resource.resolveTreeInsert(key, target, position, false, data => {
+      return transaction(change, () => {
+        return resource.tableMove(data);
+      });
+    });
+  }
+
+  /**
+   * @param {CopiedChange} change
+   * @return {Promise<void>}
+   */
+  applyCopy(change) {
+    const resource = INDEXEDDB_RESOURCES[change.table];
+    if (!resource || !resource.tableCopy) {
+      return Promise.resolve();
+    }
+
+    const { key, target, position, from_key } = change;
+    // copying takes the ID of the node to copy, so we use `from_key`
+    return resource.resolveTreeInsert(from_key, target, position, true, data => {
+      return transaction(change, () => {
+        // Update the ID on the payload to match the received change, since isCreate=true
+        // would generate new IDs
+        data.payload.id = key;
+        return resource.tableCopy(data);
+      });
+    });
+  }
+
+  /**
+   * @param {PublishedChange} change
+   * @return {Promise<void>}
+   */
+  applyPublish(change) {
+    if (change.table !== TABLE_NAMES.CHANNEL) {
+      return Promise.resolve();
+    }
+
+    // Publish changes associate with the channel, but we open a transaction on contentnode
+    return transaction(change, TABLE_NAMES.CONTENTNODE, () => {
+      return db
+        .table(TABLE_NAMES.CONTENTNODE)
+        .where({ channel_id: change.channel_id })
+        .modify({ changed: false, published: true });
```

Comment:
This was pre-existing - but it's just occurred to me that this could mark content nodes are published that are not, if changes have been applied to them since the publish event was submitted.

Don't think we need to unpick this now, but might be good to have a follow up issue.

Reply:
Yes that makes sense. Surprising we had not thought of this yet...
https://github.com/learningequality/studio/issues/4138